apps/hash-ai-worker-ts/src/activities/flow-activities/persist-entities-action.ts

```diff
@@ -0,0 +1,179 @@
+import type { GraphApi } from "@local/hash-graph-client";
+import { getSimplifiedActionInputs } from "@local/hash-isomorphic-utils/flows/action-definitions";
+import type {
+  FailedEntityProposal,
+  PersistedEntities,
+  PersistedEntity,
+  ProposedEntityWithResolvedLinks,
+} from "@local/hash-isomorphic-utils/flows/types";
+import type { OwnedById } from "@local/hash-subgraph";
+import { StatusCode } from "@local/status";
+
+import { persistEntityAction } from "./persist-entity-action";
+import type { FlowActionActivity } from "./types";
+
+export const persistEntitiesAction: FlowActionActivity<{
+  graphApiClient: GraphApi;
+}> = async ({ inputs, graphApiClient, userAuthentication }) => {
+  const { draft, proposedEntities, webId } = getSimplifiedActionInputs({
+    inputs,
+    actionType: "persistEntities",
+  });
+
+  const ownedById = webId ?? (userAuthentication.actorId as OwnedById);
+
+  /**
+   * This assumes that there are no link entities which link to other link entities, which require being able to
+   * create multiple entities at once in a single transaction (since they refer to each other).
+   *
+   * @todo handle links pointing to other links via creating many entities at once, unblocked by H-1178
+   */
+  const entitiesWithDependenciesSortedLast = [...proposedEntities].sort(
+    (a, b) => {
+      if (
+        (a.sourceEntityLocalId && b.sourceEntityLocalId) ||
+        (!a.sourceEntityLocalId && !b.sourceEntityLocalId)
+      ) {
+        return 0;
+      }
+
+      if (a.sourceEntityLocalId) {
+        return 1;
+      }
+
+      return -1;
+    },
+  );
+
+  const failedEntitiesByLocalId: Record<string, FailedEntityProposal> = {};
+  const persistedEntitiesByLocalId: Record<string, PersistedEntity> = {};
+
+  for (const unresolvedEntity of entitiesWithDependenciesSortedLast) {
```

Comment:
Yes, it would – I've added a note in https://github.com/hashintel/hash/pull/4298/commits/b8b4cbd03ef6fc305eb197093cf1b43b1bed3f91

We can see how long these steps take, but I'd expect that the persistence step will be a tiny fraction of the inference step, and they will both scale with the number of entities inferred, so that this step is never a meaningful part of the time to results for a Flow. Doing it sequentially also has the slight advantage of staggering the requests to the API, so that in theory it has to deal with less load at once.

The ideal thing here would be to be able to batch the creation of entities, so that we can do it in a single request (H-1178)